### PR TITLE
steam: add extraEnv argument

### PR DESCRIPTION
**Pull request by Atemu**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/steam.nix [hunk 1]
```diff
@@ -29,6 +30,23 @@ in {
       '';
     };
 
+    env = mkOption {
```
Comment (bjornfor) [inline nixos/modules/programs/steam.nix:33]: Why does steam need local env vars? Why not use an existing NixOS option to set env vars globally?

Comment (Atemu) [reply nixos/modules/programs/steam.nix:33]: See the example; these are env vars you might want to set for all games but not system-wide.

**Comment by Atemu** [comment]
Bump.

File: nixos/modules/programs/steam.nix [hunk 2]
```diff
@@ -29,6 +30,23 @@ in {
...
+      type = types.attrs;
+      description = mdDoc ''
+        Environment variables to run Steam with.
+
+        Uses `lib.strings.toShellVars` internally; see the documentation for more info.
```
Comment (SuperSandro2000) [inline nixos/modules/programs/steam.nix:38]: Can we move this to apply and defaultText to make it obvious what happens?

File: pkgs/games/steam/fhsenv.nix [hunk 3]
```diff
@@ -249,6 +251,9 @@ in buildFHSUserEnv rec {
 
     ${exportLDPath}
     ${fixBootstrap}
+
+    set -o allexport # Export the following env vars
```
Comment (SuperSandro2000) [inline pkgs/games/steam/fhsenv.nix:255]: Should we turn good off afterwards again? I am not sure if it is necessary.

Comment (Atemu) [reply pkgs/games/steam/fhsenv.nix:255]: The shell script ends there with an exec, so I don't think that's necessary.

File: nixos/modules/programs/steam.nix (hunk 2 above)
Comment (Atemu) [reply nixos/modules/programs/steam.nix:38]: I'm not sure what you mean by that.

**Comment by Atemu** [comment]
Bump.

**Comment by Atemu** [comment]
If there are no objections, I'll merge this.

**Comment by Atemu** [comment]
I've refactored the steam.package option to allow for a much simpler approach than adding a new option.

**Comment by Atemu** [comment]
(The first two commits could be spun off to a separate PR.)

**Review by RaitoBezarius — APPROVED** [review]
Untested; but it will probably be tested after merge very well. :)

**Comment by Atemu** [comment]
@RaitoBezarius reviewed the new and improved `programs.steam.package` option in the issue above.

I think this is ready to be unleashed upon unstable gamers.

**Pull request merged**